### ISSUE TITLE
NCacheDistributedCache constructor is requesting NCacheSessionConfiguration but only NCacheConfiguration is setup.

### ISSUE THREAD

**Issue by DamienLaw** [Alachisoft/NCache#38]
In order to use NCache as distributed cache in ASP.NET Core, one would have the following code in 
`ConfigureService()` of `Startup.cs`.

```
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddNCacheDistributedCache(Configuration.GetSection("NCache"));
    }
}
```

Looking at the source codes of `AddNCacheDistributedCache()`, it's setting up the configuration `NCacheConfiguration`.
![image](https://user-images.githubusercontent.com/14930009/54334016-952b6f80-465f-11e9-8029-3179aafdaffc.png)

However, the constructor of `NCacheDistributedCache` is requesting for `NCacheSessionConfiguration`, and not `NCacheConfiguration`.
![image](https://user-images.githubusercontent.com/14930009/54334117-edfb0800-465f-11e9-95c0-929cf09604ad.png)

Since the configuration `NCacheSessionConfiguration` was never setup, I always receive the exception **A valid Cache ID has not been specified.**
Is there something that I'm missing?

**Comment by Brad-NCache** [comment]
Hi Damien,

To use the NCache IDistributedCache provider for ASP.NET Core, the AddNCacheDistributedCache extension method provided by NCache on IServiceCollection allows configuration settings for storing sessions. The CacheName field is one of the configuration settings which is a required field. 
If you do not specify the cache ID of the cache tasked with storing sessions, you will face an error as is the scenario you are facing. 
There are two methods of passing in the configuration settings for ASP.NET Core session storage:

1. Inline In Your Application Using IOptions:
       Example: 
      
       services.AddMvc();
       services.AddNCacheDistributedCache(configuration =>
      {
               configuration.CacheName = "myCache",
              // other options
      });

2. Specifying configurations in Appsettings.json:

Sample Appsettings.json file:
{
  "AppSettings": {
    "SiteTitle": "ASP.NET MVC Music Store",
    "CacheDbResults": true,
  },

  "NCacheSettings": {
    "CacheName": "myDistributedCache",
    "EnableLogs": "True",
    "RequestTimeout": "90"
  },
}

In Application:

services.AddMvc();
services.AddNCacheDistributedCache(Configuration.GetSection("NCacheSettings"));

You can find more information about ASP.NET Core NCache IDistributed Cache Provider and the different configuration settings you can apply for session storage by using the following link:

http://www.alachisoft.com/resources/docs/ncache/prog-guide/aspnetcore-sessions-ncache-idistributedcache-provider.html

**Comment by Brad-NCache** [comment]
Hi Damien,

Thank you for pointing this out to us and helping us narrow down the cause. I have reproduced the issue and have confirmed with Engineering that it is a bug. You are correct in that the constructor of **NCacheDistributedCache** should take as a parameter **IOptions<NCacheConfiguration>** instead of **IOptions<NCacheSessionConfiguration>**. 

For your work, there are two possible workarounds:

- One method is to use **services.AddNCacheSession(Configuration.GetSection("NCache")** instead of **services.AddSession()** after **services.AddNCacheDistributedCache(Configuration.GetSection("NCache")** in the **Startup.cs** file as such (Note that I have commented out overloaded extension methods to programmatically include configuration information that you can also use):
```
public void ConfigureServices(IServiceCollection services)     
{
            services.AddMvc();
           
           // The following code adds the NCache IDistributedCache provider
            **services.AddNCacheDistributedCache(Configuration.GetSection("NCacheSettings"));**

            // Alternate extension method
            //services.AddNCacheDistributedCache(configuration =>
            //{
            //    configuration.CacheName = "myPartitionedCache";
            //});
           
           // The following code adds the NCache ASP.NET Core session provider
            **services.AddNCacheSession(Configuration.GetSection("NCacheSettings"));**

            // Alternate extension method 
            //services.AddNCacheSession(configuration =>
            //{
            //    configuration.CacheName = "myPartitionedCache"; 
            //});
        }  
```    
       After this, in the **Configure** method, also within the **Startup.cs**  file, you can enable NCache functionality into the request pipeline using the **UseNCacheSession** extension method as shown in the following example:
```
public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
{
      // Some code
      
       app.UseStaticFiles();

       **app.UseNCacheSession();**

       app.UseMvc(routes =>
           {
                // routes
            });
}

```

- The second possible workaround is to use the NCache source code from GitHub, and use the source code in the **NCache-master/Integration/NCacheSessionServices**  file path. Go to **NCacheDistributedCache.cs** file, correct the constructor as shown below and then build the solution. 

```
public NCacheDistributedCache(IOptions<**NCacheConfiguration**> options)
{
            // Keep code the same here
}
```

You can then use the generated DLLs in your project. 

By using either of the two workarounds mentioned above, you can then provide the NCache IDistributedCache implementation through dependency injection into any constructor in the app such as a controller class constructor.

The following link contains more information on the different ways NCache can support ASP.NET Core:

http://www.alachisoft.com/resources/docs/ncache/prog-guide/session-storage-aspnet-core.html

I hope this resolves your issues.

**Comment by DamienLaw** [comment]
This has been resolved in the latest version 5.0. Thank you.
